### ISSUE TITLE
Cannot stream with the "on_publish" directive

### ISSUE THREAD

**Issue by nitsuj102284** [sergey-dryabzhinsky/nginx-rtmp-module#230]
I have Nginx RTMP Module compiled from Arut's master repository, but DASH streaming works in very little players that I have tested. So I searched, and someone in another thread recommended the branch from Sergey. I setup a new server instance and installed from the repository, and it DASH seems to work much better, however I have another problem.

With the repository, I can stream to the server without the "on_publish" directive, but when I add the "on_publish" directive, pointing to the same file as I do on the server compiled from Arut, I cannot publish from OBS. There is generic error in OBS that it cannot publish, but there is no error in Nginx logs.

Any idea? Is the "on_publish" directive a problem with this branch?

Thanks!

**Comment by ethernetweb** [comment]
Have never had any problems making the on_publish/done directives function as intended. Perhaps share what it is that you're linking to for the directive?

**Comment by nitsuj102284** [comment]
The only other time (with Arut's repository) that I had an issue was when I had my php route accepting 'get' requests, but Nginx RTMP Module was sending post request. I use Laravel, and actually found the error in the Laravel logs, that the HTTP method (get) was not allowed.

In this case, I am not getting any error message in the logs, Nginx or Laravel. Basically, I am linking to http://api.example.com/v1/stream/publish, which is a PHP file that takes the stream key (I use GUID, prefixed by the stream ID; e.g. 999-bb0b4535-b745-4516-b94d-796f025e64b4, where '999' is the stream id).

From there, I check the database to see if the stream key matches the stream id, check to see that the stream id does not have an active stream already, create and register a stream session id, and then redirect stream to a different app (on another port) with the stream name changed from the stream key that was provided to the stream session id  (e.g. 054CW0JHLttu).

I don't think that it is an issue with my PHP script, because it works with Arut's Master branch, and, just to bypass any redirect or other issue in the script with this branch, I tested with simply returning a '200' response from the PHP script, and still have the same result.

The nginx.conf file is SCP directly from the working server, so it is the same. Here is the RTMP section:

`#rtmp servers ---->
rtmp {
    
    #:1936 (playback) ---->
    server {

        listen 1936;
        chunk_size 4096;


        #vod applications ---->

        # ~~~~ /vod
        application vod {

            play http://vod-storage.example.com/vod;

        }

        #<---- end vod applications


        #live applications ---->

        # ~~~~ /live
        application live {

            live on;

            hls on;
            hls_path /tmp/hls;

            dash on;
            dash_path /tmp/dash;

            on_play http://api.example.com/v1/stream/pull/rtmp/var;
            
        }

        # ~~~~ /

        #<---- end live applications


    }
    #<---- end :1936 (playback)




    #:1935 (publish) ---->
    server {

        listen 1935;
        chunk_size 4096;


        #live applications ---->

        # ~~~~ /live
        application live {

            live on;

            on_publish http://api.example.com/v1/stream/publish;

        }

        # ~~~~ /

        #<---- end live applications

    }
    #<---- end :1935 (publish)

}
#<---- end rtmp servers`

**Comment by ethernetweb** [comment]
I was previously using PHP for this same purpose with this branch but recently switched to node js - however, I still have my old PHP script which sends the response back as a HEADER.
example: header("HTTP/1.0 ","200 OK");

I was also sending: flush(); ob_flush();
after the header to ensure that the output buffer was force flushed.

Try that and see if it helps. (Also, as I've always seen, are you using the dev branch of this release?)

**Comment by nitsuj102284** [comment]
Yes, I am using the dev branch from Sergey. I noticed some releases that have to do with "notify", so I tried to install a couple of earlier releases (although probably my DASH fix is in the latest release). Anyway, two of the earlier releases error out on install.

My test to simplify was to seen the '200 OK' header, but had the same result. I have not tried flush() or ob_flush.

**Comment by ethernetweb** [comment]
Well, the on_publish directive does work - have been using it for years. I suppose the next step would be to ensure that your PHP script is actually being called. I would suggest outputting the $_POST array to a log file.

**Comment by nitsuj102284** [comment]
I have been using on_publish for probably the past two years, and it works for me too. I just don't understand why everything is the same EXCEPT for the module branch that I use, and I get two different results.

From the Nginx access.log (I've repaced the IP with RTMP branch:
`
[Sergey RTMP] - - [05/Apr/2017:20:19:20 +0000] "POST /v1/stream/publish HTTP/1.0" 403 16428 "-" "-"
[Arut RTMP] - - [05/Apr/2017:20:22:52 +0000] "POST /v1/stream/publish HTTP/1.0" 302 408 "-" "-"
`

So the Nginx RTMP module from Sergey is getting a '403' error.

**Comment by ethernetweb** [comment]
Can you load the script from your web browser? It doesn't make any sense to be 302 with arut and 403 with sergey. Are you using the exact same file in the same location? Looks to me like a web server permissions issue.

**Comment by nitsuj102284** [comment]
OK... smh. New question. I promise that I bypassed my code and returned '200' response at the top before. But... I commented out my code that requires that requires that the 'tcurl' parameter be from the "publish.example.com" domain (because I am using the same server instances to deliver publish, play, etc, vod, etc.). For my testing, I am using IP addresses, but have been sending the 'tcurl' parameter attached to my stream key query string.

This has worked without a problem on Arut's Master branch, and while I am sitting here typing this reply, the answer came flooding in. While I was looking thru the issues on Sergey's repository/last night, I read where someone suggested that the parameters be reordered so that Nginx-RTMP-Module-set params could not be overwritten. That is exactly what was my issue.

I am both overjoyed and mad at myself for having beat my head against the wall since last night. Headache.

**Comment by nitsuj102284** [comment]
Here is the code that blocked it in my PHP script:

`
(explode('.', parse_url($args['tcurl'])['host'])[0] != 'publish')
		? abort(403, 'Invalid publish domain.')
		: null;
`

**Comment by ethernetweb** [comment]
So you resolved the issue? I see you say "New question" but I can't seem to discern what the question is. :)

**Comment by nitsuj102284** [comment]
Yeah. I was thinking while I was typing. At the beginning of my reply, I had a question:

> Why isn't it evaluating my query param "tcurl"

By the time that I got to the end of that paragraph, the dawn had come. :)

Just to clarify, My stream key looks like this:
`
999-bb0b4535-b745-4516-b94d-796f025e64b4?tcurl=rtmp%3A%2F%2Fpublish.example.com
`
Yes, I resolved the issue. This branch reorders the parameters, so my custom 'tcurl' parameter does not overwrite Nginx RTMP Module 'tcurl' parameter. But I know what to do. I will look for a truly custom parameter, something like 'testmode=1', else look for 'tcurl' parameter set by Nginx RTMP Module.

I really appreciate your response and help! Your comments led me to look at the access log and evaluate my PHP script one more time.

**Comment by ethernetweb** [comment]
Well, if it helps, I pass a key value as a query string parameter and match that to the permitted value and stream name. IE: my publish url is like this:   rtmp://host:port/app/stream?key=flkewjklsjdlkfjskljfd

**Comment by misiek08** [comment]
Maybe try with tcUrl (with uppercase "U"). Maybe that's the only problem
here :)

2017-04-05 22:57 GMT+02:00 nitsuj102284 <notifications@github.com>:
